File: <100/023/023.hs
import Data.Numbers.Primes ( primeFactors )
import Data.List ( group )
import Data.List.Unique ( sortUniq )


main :: IO ()
main = print $ total - multi
  where
    total     = upperLimit * (upperLimit+1) `div` 2
    multi     = sum . sortUniq $ [n+m | n <- abundants, m <- abundants, n+m <= upperLimit]
    abundants = [n | n <- [1..upperLimit], isAbundant n]


isAbundant :: Int -> Bool
isAbundant n = properSum n > n


properSum :: Int -> Int
properSum n = divisorsSum n - n


divisorsSum :: Int -> Int
divisorsSum = product . mPrimeSum . primeFactors'


-- if given number is 72 = 2^3 * 3^2, this function returns [ [2, 2, 2], [3, 3] ]
primeFactors' :: Int -> [[Int]]
primeFactors' = group . primeFactors


-- if given [[Int]] is [ [2, 2, 2], [3, 3] ], this function returns [15, 13]
mPrimeSum :: [[Int]] -> [Int]
mPrimeSum = map primeSum


-- [p, ... , p] with length n represents p^n. primeSum calculates \sum_{k=0}^n p^k
primeSum :: [Int] -> Int
primeSum [] = 0
primeSum xs =  (x^(l+1) - 1) `div` (x - 1)
    where l = length xs
          x = head xs


upperLimit :: Int
upperLimit = 28123

File: <100/023/023.jl
using Primes

"returns ∑_{k=0}^up_to r^k = 1 + r + r^2 + ⋯ + r^up_to"
function geometric_sum(r :: Integer, up_to :: Integer) :: Integer
  if r == 1
    return up_to + 1
  end
  return (r^(up_to+1) - 1) ÷ (r - 1)
end


function divisors_sum(n :: Integer) :: Integer
  pfs = factor(n)
  l = [geometric_sum(p[1], p[2]) for p ∈ pfs]
  return prod(l)
end


function proper_divisors_sum(n :: Integer) :: Integer
  return divisors_sum(n) - n
end


function is_abundant(n :: Integer) :: Bool
  return n < proper_divisors_sum(n)
end